sql/core/src/test/scala/org/apache/spark/sql/test/SharedSparkSessionWithFreshCopy.scala

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.test
+
+import org.scalatest.Suite
+
+// A variant of SharedSparkSession. It creates a fresh copy of the spark session before each test.
+trait SharedSparkSessionWithFreshCopy extends SharedSparkSession { self: Suite =>
```

Comment:
What is the reason for creating a new trait, instead of changing the `beforeEach` method in `SharedSparkSession`? Maybe a lot of tests are broken if we change `SharedSparkSession`?

Reply:
We have `SQLTestData`, which creates DataFrames and the DataFrames are shared between all the tests in one suite. A DataFrame is associated with one `SparkSession` instance, but we create a new `SparkSession` instance for each test. If we change the config in one test, the DataFrames won't get affected.

That said, to apply `SharedSparkSessionWithFreshCopy`, the test suite should either create DataFrames in each test, or create temp views. So we need to apply it incrementally.